Add a custom theme dialog

jeremypw opened the pull request:
Pulled from #552 
Fixes #418 

Potentially address #39 
Potentially addresses #71
Potentially address #501 
Partially addresses #507 
Potentially address #632 

Additional work:
 * [x] Fix appearance of dialog under RTL
 * [x] Add a reset to default palette button
 * [x] Name variables according to linked nominal color
 * [x] Add labels for light colors
 * [x] Tweak layout and add some margins  
 * [x] Persist user custom theme after switching to a builtin theme
 * [ ] Import a theme

jeremypw commented:
@danrabbit I would appreciate your comments on this. I added the extra labels as it didn't seem obvious to a new user why there were two very similar color buttons with only one label.

The main flaw with this is that it is not possible to save a custom theme permanently - if you switch back to a builtin theme it overwrites your custom theme and you cannot return to it without rebuilding it.  It might be better to fix that in a new PR as there is more than one possible solution - e.g. a new setting or a file in the user home.

jeremypw commented:
I see from the prior art in #418 that no-one else thinks the extra labels are necessary so I'll remove if required.

jeremypw commented:
There are also some possible extra refinements like using palette colors for foreground/background but that can be done in another PR.

danrabbit commented:
I think we should solve the issue of persisting the custom palette before exposing this toggle so that we don't have an immediately broken feature. We should probably make sure that the `palette` setting is only used to store values for the custom theme and that Terminal is loading the palette for built-in themes from the Theme class and not from gsettings

jeremypw commented:
Yeh, that makes sense.

jeremypw commented:
@danrabbit  After merging master and making relevant changes, the custom theme now persists. I have moved the "Color Palette" header to before the background/foreground buttons because although the palette strictly speaking does not include these, to the average user they are just some more color settings.  I have also changed the reset button to reset the foreground/background/cursor colors for the same reason.

Can importing/saving custom themes be left for a later PR, if required?

danrabbit commented on src/Themes.vala:
```diff
@@ -40,12 +40,34 @@ public class Terminal.Themes {
+        "rgba(46, 46, 46, 0.95)", "#a5a5a5", "#839496"
+    };
+    public const int PALETTE_SIZE = 19;
+    public static string[] get_string_palette (string theme) {
```
Does this need to be public?

danrabbit commented on src/Themes.vala:
```diff
@@ -40,12 +40,34 @@ public class Terminal.Themes {
+    }
+
+    // DARK_PALETTE is used as a fallback when custom palette colors are invalid
+    public const string[] DARK_PALETTE = {
```
Since we have `get_string_palette` now, should this not be a constant and have `get_rgba_palette` use `get_string_palette (prefer_dark ? DARK : LIGHT)` for its fallback?

jeremypw replied:
This is possible only if we replace the *whole* of the palette if *any* of the custom colors is invalid.  Is that OK? At the moment, we only replace individual colors.

danrabbit commented on src/Dialogs/ColorPreferencesDialog.vala:
```diff
@@ -0,0 +1,353 @@
+/*
+* Copyright (c) 2020 elementary, Inc. (https://elementary.io)
```
```suggestion
* Copyright 2022 elementary, Inc. (https://elementary.io)
```

danrabbit reviewed (changes requested):
Two small questions, but other than that totally happy with this as an initial implementation. Let's get this in and then smooth out the flow a little bit/do other features in future branches

jeremypw commented:
I have implemented your suggestion re fallback palette as it simplifies the code and the issue of possible invalid colors in the settings (presumably due to manual editing) is unlikely.  We now fallback to a default palette if *any* of the custom colors are invalid rather than replacing only the invalid color.

danrabbit commented on src/Themes.vala:
```diff
@@ -40,12 +41,27 @@ public class Terminal.Themes {
+            var new_color = Gdk.RGBA ();
+            // If custom palette invalid use a fallback one
+            if (!new_color.parse (string_palette[i])) {
+                critical ("Color %i '%s' is not valid - using fallback palette", i, string_palette[i]);
+                return get_rgba_palette (
+                    Application.settings.get_boolean ("prefer-dark-style") ? DARK : LIGHT
+                );
```
Couldn't we do this?

```suggestion
                critical ("Color %i '%s' is not valid - replacing with default", i, string_palette[i]);
                settings_valid = false;

                var fallback_palette = get_string_palette (
                    Application.settings.get_boolean ("prefer-dark-style") ? DARK : LIGHT
                );
                string_palette[i] = fallback_palette[i];
                new_color.parse (fallback_palette[i]);
```

danrabbit commented on src/Themes.vala:
```diff
@@ -63,37 +79,30 @@ public class Terminal.Themes {
-            rgba_palette[i] = new_color;
-        }
-
-        if (!settings_valid) {
-            /* Remove invalid colors from setting */
-            Application.settings.set_string ("palette", string.joinv (":", string_palette));
-        }
```
I don't think this was meant to be removed. This overwrites the setting with the sanitized palette if we detect that any of the values are invalid

jeremypw replied:
Reinstated

danrabbit reviewed (changes requested):
One suggestion and a comment

jeremypw commented:
Ok, so you want to keep correcting individual colors; I'll revert to that. I just thought that this pretty unlikely to be required in practice so simplified it. I guess if we implementing importing customs themes it raises the risk of invalid colors being imported.

danrabbit reviewed (approved):
Yeah it's just a little bit less destructive. This looks good!
